src/cblas.rs

```diff
@@ -2,382 +2,774 @@
-    pub fn caxpy<T>(N: i32, alpha: &[T], x: &[T], incx: i32, y: &mut [T], incy: i32) {
+    pub fn caxpy<T: Vector<Complex<f32>> + ?Sized>(alpha: &Complex<f32>, x: &T, y: &mut T) {
+        check_equal_len(x, y).expect("Vectors `x` and `y` must have the same length");
         unsafe {
             sys::cblas_caxpy(
-                N,
-                alpha.as_ptr() as *const _,
-                x.as_ptr() as *const _,
-                incx,
-                y.as_mut_ptr() as *mut _,
-                incy,
+                len(x),
+                alpha as *const Complex<f32> as *const _,
+                as_ptr(x) as *const _,
+                stride(x),
+                as_mut_ptr(y) as *mut _,
+                stride(y),
             )
         }
     }
 
+    #[cfg(feature = "complex")]
+    /// Swap vectors `x` and `y`.
     #[doc(alias = "cblas_zswap")]
-    pub fn zswap<T>(N: i32, x: &mut [T], incx: i32, y: &mut [T], incy: i32) {
+    pub fn zswap<T: Vector<Complex<f64>> + ?Sized>(x: &mut T, y: &mut T) {
+        check_equal_len(x, y).expect("Vectors `x` and `y` must have the same length");
         unsafe {
-            sys::cblas_zswap(
-                N,
-                x.as_mut_ptr() as *mut _,
-                incx,
-                y.as_mut_ptr() as *mut _,
-                incy,
+            ::sys::cblas_zswap(
+                len(x),
+                as_mut_ptr(x) as *mut _,
+                stride(x),
+                as_mut_ptr(y) as *mut _,
+                stride(y),
             )
         }
     }
 
+    #[cfg(feature = "complex")]
+    /// Copy the content of `x` into `y`.
     #[doc(alias = "cblas_zcopy")]
-    pub fn zcopy<T>(N: i32, x: &[T], incx: i32, y: &mut [T], incy: i32) {
+    pub fn zcopy<T: Vector<Complex<f64>> + ?Sized>(x: &T, y: &mut T) {
+        check_equal_len(x, y).expect("Vectors `x` and `y` must have the same length");
         unsafe {
-            sys::cblas_zcopy(
-                N,
-                x.as_ptr() as *const _,
-                incx,
-                y.as_mut_ptr() as *mut _,
-                incy,
+            ::sys::cblas_zcopy(
+                len(x),
+                as_ptr(x) as *const _,
+                stride(x),
+                as_mut_ptr(y) as *mut _,
+                stride(y),
             )
         }
     }
 
+    #[cfg(feature = "complex")]
+    /// `y` := `alpha` * `x` + `y`.
     #[doc(alias = "cblas_zaxpy")]
-    pub fn zaxpy<T>(N: i32, alpha: &[T], x: &[T], incx: i32, y: &mut [T], incy: i32) {
+    pub fn zaxpy<T: Vector<Complex<f64>> + ?Sized>(alpha: &Complex<f64>, x: &T, y: &mut T) {
+        check_equal_len(x, y).expect("Vectors `x` and `y` must have the same length");
         unsafe {
-            sys::cblas_zaxpy(
-                N,
-                alpha.as_ptr() as *const _,
-                x.as_ptr() as *const _,
-                incx,
-                y.as_mut_ptr() as *mut _,
-                incy,
+            ::sys::cblas_zaxpy(
```

Comment:
I changed edition so please use `crate::` instead of `::`.

Reply:
Ah, yes, I forgot to remove them during the merge.  Will do.